### PR TITLE
Xiangqi_Lite Users APIs

### PR DESCRIPTION
**Pull request by bilal02arbisoft**
This PR introduces the users app to the Django project, providing a comprehensive user management system with robust APIs, models, serializers, URL configurations, and centralized error handling. The users app is designed to handle all user-related functionality in a modular and maintainable manner

## Summary by CodeRabbit

- **New Features**
	- Introduced a custom user model with email verification and OTP functionality.
	- Added user management views for registration, authentication, and profile management.
	- Implemented a Celery task for sending OTP emails.
	- Enhanced the Django admin interface for user management.
	- Established URL routing for user-related functionalities.
	- Added structured error handling for user-related operations.
	- Introduced signal handlers for profile picture management and player profile creation.
	- Added support for asynchronous task processing with Celery.

- **Bug Fixes**
	- Improved error handling across various user-related operations.

- **Documentation**
	- Updated README and added documentation for new features and configurations.

- **Chores**
	- Cleaned up the repository by updating `.gitignore` and configuring pre-commit hooks.
	- Established foundational project settings and configurations for deployment.

### CODE REVIEW

File: users/error_handling.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    and general functions in Django.
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+
+            return func(*args, **kwargs)
+        except ObjectDoesNotExist as e:
+
+         return Response({'error': 'Resource not found', 'details': str(e)}, status=status.HTTP_404_NOT_FOUND)
```
Comment (huwaizatahir2) [inline users/error_handling.py:25]: We can create ErrorMessages and SuccessMessages class and store all of these messages there

File: users/managers.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    @handle_exceptions
+    def create_superuser(self, email, username, password=None, **extra_fields):
+        """
+        Creates and saves a superuser with the given email, username, and password.
+        """
+        extra_fields.setdefault('is_staff', True)
+        extra_fields.setdefault('is_superuser', True)
+        if extra_fields.get('is_staff') is not True:
+
```
Comment (huwaizatahir2) [inline users/managers.py:35]: space before if not after

File: users/models.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    def __str__(self):
+
+        return self.username
+
+    def generate_otp(self):
+        """
+        Generates a one-time password (OTP) for email verification and sends it to the user's email.
+        """
+        otp = ''.join(random.choices(string.digits, k=6))
```
[the remark this answers is not in the document]
Comment (huwaizatahir2) [reply users/models.py]: @bilal02arbisoft apply this

File: users/serializers.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        Create a new user and associated profile.
+        """
+        profile_data = validated_data.pop('profile', {})
+        skill_level = validated_data.pop('skill_level', None)
+        user = CustomUser.objects.create_user(
+            email=validated_data['email'],
+            username=validated_data['username'],
+            password=validated_data['password']
+        )
+        post_save.send(sender=user.__class__, instance=user, created=True, manual=True, skill_level=skill_level)
```
[the remark this answers is not in the document]
Comment (huwaizatahir2) [reply users/serializers.py:64]: @bilal02arbisoft why did you send save signal manually?

File: users/serializers.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    def validate(self, data):
+        """
+        Validate the old and new passwords.
+        """
+        old_password = data.get('old_password')
+        new_password = data.get('new_password')
+        user = self.context.get('user')
+        if not user.check_password(old_password):
+
```
Comment (huwaizatahir2) [inline users/serializers.py:111]: fix space issues everywhere

File: users/views.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    Base API view that includes default permission classes and error handling.
+    All views inheriting from this class will automatically apply the error handling decorator.
+    """
+    permission_classes = [IsAuthenticated]
+
+class UsersListView(BaseAPIView):
+    """
+    View to retrieve a list of all users. Requires authentication.
+    """
+    @method_decorator(cache_page(60 * 15))
```
Comment (huwaizatahir2) [inline users/views.py]: use constant for caching

File: users/views.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+from users.models import CustomUser, Profile
+from users.serializers import CustomUserSerializer, PasswordChangeSerializer, RequestOtpSerializer, VerifyOtpSerializer
+
+
+class BaseAPIView(APIView):
+    """
+    Base API view that includes default permission classes and error handling.
+    All views inheriting from this class will automatically apply the error handling decorator.
+    """
+    permission_classes = [IsAuthenticated]
```
Comment (huwaizatahir2) [inline users/views.py]: permission_classes = [IsAuthenticated, ]

File: users/views.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+class CustomTokenRefreshView(TokenRefreshView, BaseAPIView):
+    """
+    Custom view to handle refreshing JWT tokens using refresh token.
+    """
+    permission_classes = []
+
+    @handle_exceptions
+    def post(self, request, *args, **kwargs):
+        refresh_token = request.COOKIES.get('refresh_token')
+        request.data['refresh'] = refresh_token
```
[the remark this answers is not in the document]
Comment (huwaizatahir2) [reply users/views.py]: @bilal02arbisoft apply this

File: users/views.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    """
+    permission_classes = []
+
+    @handle_exceptions
+    def handle_not_found(self, request, *args, **kwargs):
+        response_data = {'message': 'URL Not found'}
+
+        return Response(response_data, status=status.HTTP_404_NOT_FOUND)
+
+    get = handle_not_found
```
Comment (huwaizatahir2) [inline users/views.py:238]: use Django default functionality for this and this needs improvement

**Review by huwaizatahir2 — CHANGES_REQUESTED** [review]

File: users/views.py (hunk 7 above)
Comment (bilal02arbisoft) [reply users/views.py]: done

File: users/views.py (hunk 6 above)
Comment (bilal02arbisoft) [reply users/views.py]: done

File: users/serializers.py (hunk 5 above)
Comment (bilal02arbisoft) [reply users/serializers.py:111]: done

**Pull request merged**